fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanInstallmentDelinquencyTagRemovalServiceImpl.java

```diff
@@ -0,0 +1,47 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.portfolio.loanaccount.service;
+
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.apache.fineract.portfolio.delinquency.domain.LoanInstallmentDelinquencyTag;
+import org.apache.fineract.portfolio.delinquency.domain.LoanInstallmentDelinquencyTagRepository;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class LoanInstallmentDelinquencyTagRemovalServiceImpl implements LoanInstallmentDelinquencyTagRemovalService {
+
+    private final LoanInstallmentDelinquencyTagRepository loanInstallmentDelinquencyTagRepository;
+
+    @Override
+    public void removeDelinquencyTagsForNonExistingInstallments(Long loanId) {
```

Comment:
i am not fully sure why we need this:
- Either you are deleting all of the tags and recalculate all of them and store
or
- You are removing which is not delinquent anymore / not existing (when you calculated the new entries, you have all the information to remove the "old" ones)

Reply:
Moved this logic to delinquency calculation service.